reqactor/src/backend.rs

```diff
@@ -0,0 +1,501 @@
+                    tracing::error!(
+                        "Actor Backend failed to get status of cancel-action {request_key}: {err:?}"
+                    );
+                    Err(err)
+                }
+            },
+        }
+    }
+
+    // Check the request status and then move on to the next step accordingly.
+    async fn handle_internal_signal(&mut self, request_key: RequestKey) {
+        match self.pool.get(&request_key) {
+            Ok(Some((request_entity, status))) => match status.status() {
+                Status::Registered => match request_entity {
+                    RequestEntity::SingleProof(entity) => {
+                        tracing::info!("Actor Backend received internal signal {request_key}, status: {status}, proving single proof");
+                        self.prove_single(request_key.clone(), entity).await;
+                        self.ensure_internal_signal(request_key).await;
+                    }
+                    RequestEntity::Aggregation(entity) => {
+                        tracing::info!("Actor Backend received internal signal {request_key}, status: {status}, proving aggregation proof");
+                        self.prove_aggregation(request_key.clone(), entity).await;
+                        self.ensure_internal_signal(request_key).await;
+                    }
+                },
+                Status::WorkInProgress => {
+                    // Wait for proving completion
+                    tracing::info!(
+                        "Actor Backend checks a work-in-progress request {request_key}, elapsed: {elapsed:?}",
+                        elapsed = chrono::Utc::now() - status.timestamp(),
+                    );
+                    self.ensure_internal_signal_after(request_key, Duration::from_secs(3))
+                        .await;
+                }
+                Status::Success { .. } | Status::Cancelled { .. } | Status::Failed { .. } => {
+                    tracing::info!("Actor Backend received internal signal {request_key}, status: {status}, done");
+                }
+            },
+            Ok(None) => {
+                tracing::warn!(
+                    "Actor Backend received internal signal {request_key}, but it is not in pool, skipping"
+                );
+            }
+            Err(err) => {
+                // Fault tolerance: re-enqueue the internal signal after 3 seconds
+                tracing::warn!(
+                    "Actor Backend failed to get status of internal signal {request_key}: {err:?}, performing fault tolerance and retrying later"
+                );
+                self.ensure_internal_signal_after(request_key, Duration::from_secs(3))
+                    .await;
+            }
+        }
+    }
+
+    // Ensure signal the request key to the internal channel.
+    //
+    // Note that this function will retry sending the signal until success.
+    async fn ensure_internal_signal(&mut self, request_key: RequestKey) {
```

Comment:
should this ensure_signal has expected status/response?

Reply:
No response currently. It's just a "message bus".